packages/core/src/types.ts

```diff
@@ -0,0 +1,371 @@
+// Config
+//------------------------------------------------------------------------------
+
+export type RuleSeverity = 0 | 1 | 2 | "off" | "warn" | "error";
+
+export type RuleConfig = RuleSeverity | [RuleSeverity, ...any[]];
+
+export interface RulesConfig {
+	[key: string]: RuleConfig;
+}
+
+//------------------------------------------------------------------------------
+// Languages
+//------------------------------------------------------------------------------
+
+/**
+ * Represents a plugin language.
+ */
+export interface Language {
+	/**
+	 * Indicates how ESLint should read the file.
+	 */
+	fileType: "text"; // future will also support "binary"
+
+	/**
+	 * First line number returned from the parser (text mode only).
+	 */
+	lineStart: 0 | 1;
+
+	/**
+	 * First column number returned from the parser (text mode only).
+	 */
+	columnStart: 0 | 1;
+
+	/**
+	 * The property to read the node type from. Used in selector querying.
+	 */
+	nodeTypeKey: string;
+
+	/**
+	 * Validates languageOptions for this language.
+	 */
+	validateLanguageOptions(languageOptions: LanguageOptions): void;
+
+	/**
+	 * Helper for esquery that allows languages to match nodes against
+	 * class. esquery currently has classes like `function` that will
+	 * match all the various function nodes. This method allows languages
+	 * to implement similar shorthands.
+	 */
+	matchesSelectorClass?(
+		className: string,
+		node: SyntaxElement,
+		ancestry: Array<SyntaxElement>,
+	): boolean;
+
+	/**
+	 * Parses the given file input into its component parts. This file should not
+	 * throws errors for parsing errors but rather should return any parsing
+	 * errors as parse of the ParseResult object.
+	 */
+	parse(file: File, context: LanguageContext): ParseResult; // Future: | Promise<ParseResult>;
+
+	/**
+	 * Creates SourceCode object that ESLint uses to work with a file.
+	 */
+	createSourceCode(
+		file: File,
+		input: ParseResult,
+		context: LanguageContext,
+	): SourceCode; // Future: | Promise<SourceCode>;
+}
+
+/**
+ * Plugin-defined options for the language.
+ */
+export type LanguageOptions = Record<string, unknown>;
+
+/**
+ * The context object that is passed to the language plugin methods.
+ */
+export interface LanguageContext {
+	languageOptions: LanguageOptions;
+}
+
+/**
+ * Represents a file read by ESLint.
+ */
+export interface File {
+	/**
+	 * The path that ESLint uses for this file. May be a virtual path
+	 * if it was returned by a processor.
+	 */
+	path: string;
+
+	/**
+	 * The path to the file on disk. This always maps directly to a file
+	 * regardless of whether it was returned from a processor.
+	 */
+	physicalPath: string;
+
+	/**
+	 * Indicates if the original source contained a byte-order marker.
+	 * ESLint strips the BOM from the `body`, but this info is needed
+	 * to correctly apply autofixing.
+	 */
+	bom: boolean;
+
+	/**
+	 * The body of the file to parse.
+	 */
+	body: string | ArrayBuffer;
```

Comment:
In ESLint, this is [`string | Uint8Array`](https://github.com/eslint/eslint/blob/e2b16e2b72606162dce3d804bc80186b6c5ec0f9/lib/linter/vfile.js#L76)

Reply:
Another good catch. :+1: